src/sv-pipeline/04_variant_resolution/scripts/VCFParser.java

```diff
@@ -0,0 +1,616 @@
+            final FileInputStream fis = new FileInputStream(pathName);
+            this.is = pathName.endsWith(GZ) ? new GZIPInputStream(new BufferedInputStream(fis)) : fis;
+        } catch ( final IOException ioe ) {
+            throw new MalformedVCFException("can't open " + pathName, ioe);
+        }
+        if ( !readBuffer() ) {
+            throw new MalformedVCFException(pathName + " is empty");
+        }
+    }
+
+    public void close() {
+        try {
+            is.close();
+        } catch ( final IOException ioe ) {
+            throw new MalformedVCFException("can't close " + pathName, ioe);
+        }
+    }
+
+    /** process lines beginning with '#' as metadata */
+    public boolean hasMetadata() {
+        if ( !bufferIterator.hasNext() ) {
+            if ( !readBuffer() ) {
+                return false;
+            }
+        }
+        return bufferIterator.peek() == '#';
+    }
+
+    public Metadata nextMetaData() {
+        if ( !hasMetadata() ) {
+            throw new NoSuchElementException();
+        }
+        bufferIterator.skip();
+        needData();
+        // are we about to parse the column header?
+        // it's the only metadata line that doesn't start with "##" but goes:
+        // #CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT   (sample names)
+        if ( bufferIterator.peek() != '#' ) {
+            return new Columns(captureColumns());
+        }
+        bufferIterator.skip();
+        // get the key part of the metadata, e.g., INFO in ##INFO=, or contig in ##contig=
+        final ByteSequence key = capture('=');
+        needData();
+        // metadata is just ##key=value, but some keys (like INFO) have multiple attributes
+        // as a value.  check to see if we need to tokenize multiple values.
+        // multiple values are surrounded by '<' and '>'
+        if ( bufferIterator.peek() != '<' ) {
+            // nope, simple value.  just grab the rest of the line
+            final ByteSequence value = capture('\n');
+            return new KeyValue(key, value);
+        }
+        bufferIterator.skip();
+        // yup.  multiple values. tokenize them.
+        return new KeyAttributes(key, captureAttributes());
+    }
+
+    /** once we've had hasMetadata return false (line doesn't start with '#')
+     * we're ready to read records.
+     * if there are more lines to read, they're all data records. */
+    public boolean hasRecord() {
+        if ( !bufferIterator.hasNext() ) {
+            return readBuffer();
+        }
+        return true;
+    }
+
+    /** tokenize the next line as columns, and stuff them into a Record */
+    public Record nextRecord() {
+        if ( !hasRecord() ) {
+            throw new NoSuchElementException();
+        }
+        if ( bufferIterator.peek() == '#' ) {
+            throw new IllegalStateException("can't read records until metadata is exhausted");
+        }
+        return new Record(captureColumns());
```

Comment:
I think it would be problematic.  I'd have to allocate and reallocate bigger and bigger byte buffers to save all the trailing fields.  Better to just grab them out of the current buffer (making a new ByteSequence involves no allocation or copying).